### PR TITLE
Add SPIR-V language with SPIRV-Tools compilers

### PR DESCRIPTION
**Pull request by spencer-lunarg**
for https://github.com/compiler-explorer/compiler-explorer/issues/2331

Hey, so I am currently trying to improve the SPIR-V ecosystem, in-charge of the [SPIR-V Guide](https://github.com/KhronosGroup/SPIRV-Guide), and part of the SPIR-V Working Group. We talked at our last Face-2-Face gathering about getting more support for GPU tooling in something like Compiler Explorer because you have all made this such an amazing tool!

I have enjoyed working in this codebase and happy to help keep improving the SPIR-V as people find issues or want enhancements (I was already about to spend time rebuilding Compiler Explorer, so happy to donate my time to a single tool's effort)

(For those who haven't spent last few years staring at SPIR-V)

- SPIR-V is a an IR (very much like LLVM IR) but for GPUs
- SPIR-V can be used for Graphics and Compute
- Compute:
    - Can be tied to things like clang, llvm, mlir, OpenCL, etc
    - One could take `spirv` as a language and compile it to something like x86
      - (Future PRs for this realm of things)
- Graphics:
    - Much harder to actually "compile" to any ISA because other graphics pipeline information is missing that will effect the final ISA
    - Microsoft has even now planned to move to [SPIR-V for DirectX soon](https://devblogs.microsoft.com/directx/directx-adopting-spir-v/)
- Both:
    - There are many tools that run on SPIR-V regardless (hence why they share the same IR)
    - Being able to do tooling in compiler explorer would be AMAZING

This PR adds `SPIR-V` as a language and a new `SPIRV-Tools` compiler as an initial compiler for it the language (since we already have the SPIRV-Tools repo being pulled in on the infra side!)

I personally use `spirv-opt` and `spirv-val` the most and have been testing this locally for a few days and feel it is ready to go. I plan to slowly add more tools (ex. `spirv-fuzz`, `spirv-cross`, `spirv-reflect`, etc)

Cheers!

(some screenshots of course)
image
image

### CODE REVIEW

**Comment by partouf** [comment]
This definitely looks good, thanks for your efforts.
I hope we can merge this somewhere next week

**Comment by mattgodbolt** [comment]
Sorry for the delay on this @spencer-lunarg - I'm taking a look at this now.

File: lib/compilers/spirv.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2018, 2021, Compiler Explorer Authors, Arm Ltd
+// Copyright (c) 2018, 2021, 2024 Compiler Explorer Authors, Arm Ltd
```
Comment (mattgodbolt) [inline lib/compilers/spirv.ts:1]: Thanks for updating this; though typically we leave the copyright alone but not really for any principled reason...

File: lib/compilers/spirv.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ import {logger} from '../logger.js';
 import {SPIRVAsmParser} from '../parsers/asm-parser-spirv.js';
 import * as utils from '../utils.js';
 
+// If you want to output SPIR-V, most likely you want SPIRVAsmParser
```
Comment (mattgodbolt) [inline lib/compilers/spirv.ts:40]: 🙇🏻

File: etc/config/spirv.amazon.properties [hunk 3]
```diff
@@ -0,0 +1,21 @@
+compilers=&spirv-opt:&spirv-val
+defaultCompiler=spirv-opt
+
+supportsBinary=false
+supportsExecute=false
+supportsAsmDocs=false
+
+group.spirv-opt.groupName=Optimizer
+group.spirv-opt.baseName=SPIRV-Tools Optimizer
+group.spirv-opt.compilers=spirv-opt
```
Comment (mattgodbolt) [inline etc/config/spirv.amazon.properties:10]: From bitter experience it's better to name these IDs with the version too; else one day when/if we add a versioned compiler it's ambiguous what version "spirv-opt" is.

Can you rename the individual `compiler.*` names to be `spirv-opt-master` and `spirv-val-master` to future proof us here? Thanks!

File: lib/compilers/spirv-tools.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    getPrimaryOutputFilename(dirPath: string, outputFilebase: string) {
+        return path.join(dirPath, `${outputFilebase}.spv`);
+    }
+
+    override getOutputFilename(dirPath: string, outputFilebase: string) {
+        return path.join(dirPath, `${outputFilebase}.spvasm`);
+    }
+
+    // Some tools (ex. spirv-opt) needs this as a single argument, so merge with "="
```
Comment (mattgodbolt) [inline lib/compilers/spirv-tools.ts:76]: Thanks for the handy comment here.

File: lib/compilers/spirv-tools.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    mergeSpirvTargetEnv(options: string[]) {
+        const index = options.indexOf('--target-env');
+        if (index !== -1) {
+            options.splice(index, 2, `--target-env=${options[index + 1]}`);
+        }
+        return options;
+    }
+
+    // Some tools (ex. spirv-val) needs this as a two argument, so unmerge with "="
```
Comment (mattgodbolt) [inline lib/compilers/spirv-tools.ts:85]: Ditto. Also what a shame! 😁

**Review by mattgodbolt — APPROVED** [review]
Looks good to me! Will merge and play about locally and then get this deployed.

File: etc/config/spirv.amazon.properties (hunk 3 above)
Comment (mattgodbolt) [reply etc/config/spirv.amazon.properties:10]: I'll make this change as it was the only minor comment and you've waited long enough!

Comment (spencer-lunarg) [reply etc/config/spirv.amazon.properties:10]: smart, saw the fix you made, looks good to me!

File: lib/compilers/spirv-tools.ts (hunk 5 above)
Comment (spencer-lunarg) [reply lib/compilers/spirv-tools.ts:85]: ya... been a low hanging task to fix and support both, one day ... lol

**Comment by mattgodbolt** [comment]
This, along with the glsl is rolling out now over the next half hour or so!

**Comment by spencer-lunarg** [comment]
> This, along with the glsl is rolling out now over the next half hour or so!

Awesome, thanks @mattgodbolt - excited to share with the greater GPU Developer ecosystem, will probably result in some feature requests for me and hope to be back, been pleasant to add to this tool

**Pull request merged**